internal/app/synchronizer/evaluator_client.go

```diff
@@ -103,56 +106,43 @@ func (ec *evaluatorClient) initialize() error {
 	return nil
 }
 
-func (ec *evaluatorClient) grpcEvaluate(ctx context.Context, proposals []*pb.Match) (results []*pb.Match, err error) {
-	resp, err := ec.grpcClient.Evaluate(ctx, &pb.EvaluateRequest{Matches: proposals})
-	if err != nil {
-		return nil, err
-	}
-
-	return resp.GetMatches(), nil
-}
-
-func (ec *evaluatorClient) httpEvaluate(ctx context.Context, proposals []*pb.Match) (results []*pb.Match, err error) {
-	proposalIDs := getMatchIds(proposals)
-	jsonProposals, err := json.Marshal(proposals)
-	if err != nil {
-		return nil, status.Errorf(codes.FailedPrecondition, "failed to marshal proposals to string for proposals %s: %s", proposalIDs, err.Error())
-	}
-
-	reqBody, err := json.Marshal(map[string]json.RawMessage{"match": jsonProposals})
-	if err != nil {
-		return nil, status.Errorf(codes.FailedPrecondition, "failed to marshal request body for proposals %s: %s", proposalIDs, err.Error())
-	}
-
-	req, err := http.NewRequest("POST", ec.baseURL+"/v1/matches:evaluate", bytes.NewBuffer(reqBody))
-	if err != nil {
-		return nil, status.Errorf(codes.FailedPrecondition, "failed to create evaluator http request for proposals %s: %s", proposalIDs, err.Error())
-	}
-
-	resp, err := ec.httpClient.Do(req.WithContext(ctx))
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to get response from evaluator for proposals %s: %s", proposalIDs, err.Error())
-	}
-
-	defer func() {
-		err = resp.Body.Close()
-		if err != nil {
-			evaluatorClientLogger.WithError(err).Warning("failed to close response body read closer")
+func (ec *evaluatorClient) grpcEvaluate(ctx context.Context, proposals []*pb.Match) ([]*pb.Match, error) {
+	stream, err := ec.grpcClient.Evaluate(ctx)
+
+	waitc := make(chan evaluatorResult)
+	defer close(waitc)
+	go func() {
+		results := []*pb.Match{}
+		for {
+			resp, recvErr := stream.Recv()
+			if recvErr == io.EOF {
+				// read done.
+				waitc <- evaluatorResult{results: results, err: nil}
+				return
+			}
+			if recvErr != nil {
+				waitc <- evaluatorResult{results: nil, err: recvErr}
+				return
+			}
+			results = append(results, resp.GetMatch())
 		}
 	}()
 
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, status.Errorf(codes.FailedPrecondition,
-			"failed to read from response body for proposals %s: %s", proposalIDs, err.Error())
+	for _, proposal := range proposals {
+		if err = stream.Send(&pb.EvaluateRequest{Match: proposal}); err != nil {
+			return nil, err
```

Comment:
If this fails, simply return? Close the stream to cancel the Evaluate call? Would you rather break?

Reply:
I think the code is correct since it performed the required no-leakage actions.
 https://github.com/grpc/grpc-go/blob/6fb38bdee17208f9e2f933614eb6d9052095be22/stream.go#L129.